src/lib/structures/uid.ml

```diff
@@ -26,38 +26,81 @@
 (**************************************************************************)
 
 module DStd = Dolmen.Std
-module DE = DStd.Expr
+module DE = Dolmen.Std.Expr
 
 type t =
   | Hstring : Hstring.t -> t
   | Dolmen : 'a DE.id -> t
+  | Term_cst : DE.term_cst -> t
+  | Ty_cst : DE.ty_cst -> t
+
+type hash = {
+  to_int : t -> int;
+  (** Return a perfect hash for the constructor. This hash is between [0]
+      and [n] where [n] is the number of constructors of the ADT. *)
+
+  of_int : int -> t;
+  (** Return the associated constructor to the integer.
+
+      @raises Invalid_argument if the integer does not correspond to
+       a constructor of this ADT. *)
+}
 
 let[@inline always] of_dolmen id = Dolmen id
```

Comment:
Do we still need this case?

Reply:
Technically, we don't need it because we only store two kind of identifiers: `ty id` and `type_fun id`.

Reply:
If there is a single other kind of identifier we store, it would be better as it would avoid an existential in the type.